hack/prepare-release.sh

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env bash
+#
+# Requirements:
+# - yq v4.x
+
+set -eo pipefail
+
+usage() {
+  echo "Helper script to prepare the docs for upcoming product releases."
+  echo
+  echo "Usage:"
+  echo "  prepare-release.sh -p <PRODUCT> -v <VERSION>"
+  echo
+  echo "Flags:"
+  echo "  -p    Product selection. One of 'kubermatic', 'kubeone', 'kubecarrier'. (env: PRODUCT)"
+  echo "  -v    Version of the upcoming release. (env: VERSION)"
+  echo "  -h    Print this help."
+  echo
+  echo "Product and Version can be passed by flag or environment variable. The flag has the higher weight."
+  echo "For the component version update the code of kkp needs to be located at ../kubermatic."
+}
+
+line() {
+  printf "| %-30s | %-30s |\n" "$1" "$2"
+}
+
+version_table() {
+  line "KKP Component" "Version"
+  line "---" "---"
+  for comp in $(find ../kubermatic/charts -name 'Chart.yaml' -printf '%h\n')
+  do
+    name=${comp#../kubermatic/charts/}
+    ver=$(yq e '.appVersion' $comp/Chart.yaml | sed "s/__KUBERMATIC_TAG__/${VERSION}.0/;s/^v//")
+    line $name $ver
+  done
+}
+
+while getopts "hp:v:" option
+do
+  case $option in
+    h)
+      usage
+      exit 0;;
+    p)
+      export PRODUCT=${OPTARG};;
+    v)
+      export VERSION=${OPTARG};;
+    \?)
+      usage
+      exit 1;;
+  esac
+done
+
+# Check requirements
+yq --version | grep -q 'yq .* version 4' || {
+  echo "yq >= 4.x is required"
+  echo "e.g. go install github.com/mikefarah/yq/v4@latest"
+  exit 1
+}
+
+# Check mandatory parameters
+[[ $VERSION ]] ||
+  (usage; exit 1)
+
+[[ $PRODUCT =~ ^(kubermatic|kubeone|kubecarrier)$ ]] ||
+  (usage; exit 1)
+
+# Update component versions and copy static images only when preparing docs for KKP release
+if [[ $PRODUCT == 'kubermatic' ]]
+then
+  tmpfile=$(mktemp)
+  head -13 content/kubermatic/master/architecture/support_policy/KKP_components_versioning/_index.en.md > $tmpfile
```

Comment:
I think this is quite brittle for two reasons:
1. It copies over the date timestamp from the previous document. This results in all future documents having the same timestamp
2. It is highly dependent that the content stays within 13 lines. So if someone were to add a line in the file, it would not work correctly anymore. Additionally the script would not break, but just produce a wrong document. This is quite dangerous as someone could easily miss it, if that person does not manually check the file every time. 

To solve this I could think of three possible solutions:

### A) Use sed to anchor on a piece of text

Either this could be binding to the table on the "| KKP Components" somehow. Or you could add a line "compatibility matrix" above the table and bind to this. Anyways the idea is basically to anchor to a piece of text and copy anything above it. The advantage is that this either works when the number of lines change or will fail if the anchoring text has changed.

And then to solve the date timestamp, you can just use a shell function to insert it.


### B) Create a gohugo archetype

In this solution you would add a "Components Versioning" archetype, and then let gohugo create your new file. Basically the archetype would contain the 12 lines. And then you simply create the new file using the `hugo new` command.

In this case, the timestamp would automatically be created by hugo.

This solution of course then has hugo as a dependency for the script.


### C) Hybrid between the two

If you don't want to use hugo, I think both ideas could also be combined. Basically create a markdown file with the 12 lines to be used as a template. And then just use a bash utility to generate the timestamp.




If you choose B) or C), please also add a markdown commend in the final doc that indicates that it is auto-generated.

Reply:
Good points, I agree this hardcoded approach will break if the upper part of the page changes. I will change it to a more dynamic approach.

---

Regarding the document creation timestamp: This holds true for all the content that is copied over. If I change this here, I actually have to set it on every single markdown document I am copying over from `master` to `<version>`. As far as I have checked docs of previous releases, this has not been done before. I can introduce it everywhere, but is it really worth it? It would also only affect the metadata of the docs for new releases.

Reply:
ah I see. I did not know that the final timestamp does not show up in the document. Then it would be fine from my side to leave it as is